src/Evolution/Systems/Cce/Actions/Psi0Matching.hpp

```diff
@@ -0,0 +1,55 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <tuple>
+#include <utility>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "Evolution/Systems/Cce/NewmanPenrose.hpp"
+#include "Evolution/Systems/Cce/PreSwshDerivatives.hpp"
+
+namespace Cce {
+namespace Actions {
+
+/*!
+ * \ingroup ActionsGroup
+ * \brief Calculate \f$\Psi_0\f$ and its radial derivative
+ * \f$\partial_\underline{\lambda}\Psi_0\f$ at the inner boundary of
+ * the CCE grid.
+ *
+ * \details This action is to be called after `Tags::BondiJ` has been updated.
+ * Note that `Tags::BondiJ` is in the asymptotically inertial frame, while the
+ * returned \f$\Psi_0\f$ and its radial derivative are in the Cauchy frame.
+ * \f$\underline \lambda\f$ is an affine parameter along the null
+ * rays generated by \f$l\f$, see Eq. (19a) of \cite Moxon2020gha.
+ */
+struct CalculatePsi0AndDerivAtInnerBoundary {
+  using const_global_cache_tags = tmpl::list<Tags::LMax>;
+
+  template <typename DbTags, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
+  static Parallel::iterable_action_return_t apply(
+      db::DataBox<DbTags>& box,
+      const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+      const Parallel::GlobalCache<Metavariables>& /*cache*/,
+      const ArrayIndex& /*array_index*/, const ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) {
+    db::mutate_apply<TransformBondiJToCauchyCoords>(make_not_null(&box));
+    db::mutate_apply<PreSwshDerivatives<Tags::Dy<Tags::BondiJCauchyView>>>(
```

Comment:
Since these are just `mutate_apply`s, and I will need effectively this exact sequence of mutators for LTS (but not in an iterable action) can you do something like

```cpp
struct CalculatePsi0AndDerivAtInnerBoundary {
  using mutators = tmpl::list<TransformBondiJToCauchyCoords,
                    PreSwshDerivatives<Tags::Dy<Tags::BondiJCauchyView>>, ...>;

  static apply(DataBox& box, ...) {
  tmpl::for_each<mutators>(
       [&box](auto mutator_v){
        using mutator = std::decay_t<decltype(mutator_v)>;
        db::mutate_apply<mutator>(make_not_null(&box));
       }
    );
  }
};
```

so that I can just grab this `tmpl::list` of mutators and use it? I think that would reduce code duplication a lot. Just squash this in with your other fixups.

Reply:
Done. Although `std::decay_t`  didn't work, so I use `typename decltype(mutator_v)::type` instead.

Reply:
Ahh yeah, I think when you use `tmpl::for_each` it wraps the `mutator_v` type in some `tmpl` object and the actual type is accessed from the `::type` alias.